### PR TITLE
Misc fixes on gpexpand and data redistribution

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
These fixes are cherry picked from the feature branch https://github.com/greenplum-db/gpdb/tree/feature_gpdb6_expand, please refer to each commit for details.

## Here are some reminders before you submit the pull request
- [x] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [x] Pass `make installcheck`
- [x] Review a PR in return to support the community

### CODE REVIEW

File: src/test/regress/sql/partial_table.sql [hunk 1]
```diff
@@ -43,6 +44,11 @@ analyze r2;
 -- regression tests
 --
 
+-- Test numsegments properity cannot be larger than the size of cluster
+create table size_sanity_check(c1 int, c2 int);
+update gp_distribution_policy set numsegments = 10 where localoid = 'size_sanity_check'::regclass;
```
Comment (ashwinstar) [inline src/test/regress/sql/partial_table.sql]: Keeping such broken table around is dangerous, lets drop it in this test.

File: src/backend/cdb/cdbcat.c [hunk 2]
```diff
@@ -331,6 +331,24 @@ GpPolicyFetch(Oid tbloid)
...
+		 * of cluster and dispatcher cannot handle this.
+		 */
+		if ((Gp_role == GP_ROLE_DISPATCH || Gp_role == GP_ROLE_EXECUTE) &&
+			policyform->numsegments > GP_POLICY_ALL_NUMSEGMENTS)
+		{
+			ReleaseSysCache(gp_policy_tuple);
+			ereport(ERROR,
+					(errcode(ERRCODE_GP_FEATURE_NOT_YET),
+					 errmsg("\"numsegments\" of table cannot be larger than the size of the cluster"),
+					 errhint("Table might be expanded after current global transaction is started, re-run the query.")));
```
Comment (ashwinstar) [inline src/backend/cdb/cdbcat.c]: I am having hard-time understanding the case, "the numsegments might be larger than
the size of cluster, eg, table is expanded after a global transaction is started". Please can you elaborate more the flow.

File: src/test/regress/sql/partial_table.sql (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/test/regress/sql/partial_table.sql]: Sure, makes sense, I'll drop it.

BTW, about the commit message, when GDD is disabled updates to tables can not be paralleled as the lock level is upgraded to `ExclusiveLock`.  Only when GDD is enabled parallel updates are allowed, depends on which segments the locks are held on there can be either local deadlock or global dead lock.

File: src/backend/cdb/cdbcat.c (hunk 2 above)
Comment (pivotal-ning-yu) [reply src/backend/cdb/cdbcat.c]: Suppose the original cluster size is 3, a transaction T is running, then:
1. expand the cluster to 4;
2. T still sees the cluster size as 3;
3. create a new table  in a new transaction, its numsegments is 4;
4. T wants to access this new table, what would happen?

To access a table whose numsegments is 4, we need a gang whose size >=4, but as T can not see the new segments, it's impossible for T to access this table correctly, so we error out here as early as possible.

File: src/test/regress/sql/partial_table.sql (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/test/regress/sql/partial_table.sql]: I checked the sql file, all the tables in this file are created in the "test_partial_table" schema, this schema is dropped at end of the file.  In fact all test files that contain partial tables have such kind of cleanup.

**Comment by pivotal-ning-yu** [comment]
@dujialun the commit pushed by you broke the compilation:
https://travis-ci.org/greenplum-db/gpdb/builds/499596031?utm_source=github_status&utm_medium=notification

Please help to correct the problem.

**Comment by dujialun** [comment]
> @dujialun the commit pushed by you broke the compilation:
> https://travis-ci.org/greenplum-db/gpdb/builds/499596031?utm_source=github_status&utm_medium=notification
> 
> Please help to correct the problem.

The declaration of GP_POLICY_ALL_NUMSEGMENTS has not retired in this branch and I can compile successfully on my local environment, why it fails on pipeline?

**Comment by pivotal-ning-yu** [comment]
Ah, looks like travis automatically merged the PR on master (of course locally) and do the compilation on that.  By doing similar thing I can trigger the compilation failure.  I'll rebase the PR on current master branch and fix the failure.

**Comment by pivotal-ning-yu** [comment]
Rebased the PR on current master and force-pushed.

**Comment by pivotal-ning-yu** [comment]
Pushed a commit to ensure all partial tables are dropped after reshuffle tests and partial table tests.

File: src/backend/cdb/cdbcat.c (hunk 2 above)
Comment (ashwinstar) [reply src/backend/cdb/cdbcat.c]: Based on that description, it seems error message and hint needs to be rephrased. Current error message is saying something is wrong with the table, which is not the case. Table is fine and created okay. Problem is this transaction / query cannot use the table, as was created after it started and has different numsegments compared to this transaction can use.

Comment (pivotal-ning-yu) [reply src/backend/cdb/cdbcat.c]: Makes sense, we'll update the error message, thanks for catching this.

Comment (pivotal-ning-yu) [reply src/backend/cdb/cdbcat.c]: Updated the error messages, now it looks like below:

```sql
select * from size_sanity_check;
ERROR:  cannot access table "size_sanity_check" in current transaction
LINE 1: select * from size_sanity_check;
                      ^
DETAIL:  New segments are concurrently added to the cluster during the execution of current transaction, the table has data on some of the new segments, but these new segments are invisible and inaccessi
HINT:  Re-run the query in a new transaction.
```

**Comment by pivotal-ning-yu** [comment]
Are there any other comments? If no then we will merge this PR tomorrow, last chance to raise your concerns or suggestions.

**Comment by pivotal-ning-yu** [comment]
Rebased with latest master to resolve conflicts.

**Comment by pivotal-ning-yu** [comment]
We'll merge it now, thanks for the review.

**Pull request merged**